traits_futures/tests/i_message_router_tests.py

```
@@ -0,0 +1,386 @@
+                sender.stop()
+
+    def test_sender_stop_before_start(self):
+        with self.started_router() as router:
+            with self.get_sender(router) as sender:
+                with self.assertRaises(RuntimeError):
+                    sender.stop()
+
+    def test_sender_stop_twice(self):
+        with self.started_router() as router:
+            with self.get_sender(router) as sender:
+                sender.start()
+                sender.stop()
+                with self.assertRaises(RuntimeError):
+                    sender.stop()
+
+    def test_sender_restart(self):
+        with self.started_router() as router:
+            with self.get_sender(router) as sender:
+                sender.start()
+                sender.stop()
+                with self.assertRaises(RuntimeError):
+                    sender.start()
+
+    @contextlib.contextmanager
+    def get_sender(self, router):
+        """
+        Provide a sender for the given router, and do cleanup afterwards.
+        """
+        sender, receiver = router.pipe()
+        try:
+            yield sender
+        finally:
+            router.close_pipe(receiver)
+
+    # Helper functions and assertions
+
+    @contextlib.contextmanager
+    def started_router(self):
+        """
+        A context manager that yields an already-started router.
+
+        Stops the router on context manager exit.
+        """
+        router = self.router_factory()
+        router.start()
+        try:
+            yield router
+        finally:
+            router.stop()
+
+    def assertEventuallyReceives(
+        self, listener, messages, *, timeout=SAFETY_TIMEOUT
+    ):
+        """
+        Assert that a given listener eventually receives the given messages.
+
+        Runs the event loop until either the expected messages are received,
+        or until timeout.
+        """
+
+        def got_expected_messages(listener):
+            return listener.messages[: len(messages)] == messages
```

I'm not sure why we don't compare `listener.messages` as a whole and only the first `len(messages)` messages. Why are we expecting more?

---

Yes, this should probably be reworked. The subtlety here is that if there's a bug and we somehow go from too few messages to too many, then we need to fail there and then instead of continuing to wait. So I think what I really want to do is not compare messages at all here: just wait until I have at least the right number of messages, and then check their contents.

---

Reworked.